File: tests/ccsp-comparisons/erlang/mtring.erl
-module(mtring).
-export([main/1, ring_element/2]).

-define(ELEMENTS, 256).

next_token(0) -> 0;
next_token(Token) -> Token + 1.

ring_element(0, _) -> 
   done;
ring_element(_, Next) ->
   receive
      Token ->
         Next ! next_token(Token),
         ring_element(Token, Next)
   end.

pass_tokens(0, _) -> done;
pass_tokens(Tokens, Next) ->
  receive
     Token ->
        Next ! (Token + 1),
        pass_tokens(Tokens - 1, Next)
  end.

recv_tokens(0, Sum) -> Sum;
recv_tokens(Tokens, Sum) ->
   receive
      Token -> recv_tokens(Tokens - 1, Sum + Token)
   end.

send_tokens(0, _) -> done;
send_tokens(Tokens, Next) ->
   Next ! Tokens,
   send_tokens(Tokens - 1, Next).

ring_inner(0, Tokens, Next) ->
   Sum = recv_tokens(Tokens, 0),
   io:fwrite("end~n"),
   io:fwrite("~b~n", [Sum]),
   Next ! 0,
   receive
      _ -> done
   end;
ring_inner(Cycles, Tokens, Next) ->
   pass_tokens(Tokens, Next),
   ring_inner(Cycles - 1, Tokens, Next).

ring(1, Cycles, Tokens, Next) ->
   Next ! 1,
   receive
      _ ->
         io:fwrite("start~n"),
         send_tokens(Tokens, Next),
         ring_inner(Cycles, Tokens, Next)
   end;
ring(N, Cycles, Tokens, Last) ->
   Next = spawn(ring, ring_element, [1, Last]),
   ring(N - 1, Cycles, Tokens, Next).

ring_root(Cycles, Tokens) ->
   ring(?ELEMENTS, Cycles, Tokens, self()).

main([A1|[A2|_]]) ->
   Cycles = list_to_integer(A1),
   Tokens = list_to_integer(A2),
   ring_root(Cycles, Tokens),
   erlang:halt().

File: tests/ccsp-comparisons/haskell/mtring.hs
import Control.Monad
import Control.Concurrent
import System.Environment
import System.IO

elements = 256

nextToken 0 = 0
nextToken t = t + 1

element this next = loop
  where loop = do
          token <- takeMVar this
	  putMVar next $! (nextToken token)
          when (token > 0) loop

passTokens 0 _ _ = do
  return Nothing
passTokens tokens this next = do
  token <- takeMVar this
  putMVar next $! (token + 1)
  passTokens (tokens - 1) this next

recvTokens 0 _ sum = do 
  return sum
recvTokens tokens this sum = do
  token <- takeMVar this
  recvTokens (tokens - 1) this (sum + token)

sendTokens 0 _ = do 
  return Nothing
sendTokens tokens next = do
  putMVar next $! tokens
  sendTokens (tokens - 1) next

rootElement 0 tokens this _ = recvTokens tokens this 0
rootElement cycles tokens this next = do
  passTokens tokens this next
  rootElement (cycles - 1) tokens this next

root :: Int -> Int -> MVar Int -> MVar Int -> IO Int
root cycles tokens this next = do
  putMVar next $! 1
  token1 <- takeMVar this
  
  hPutStrLn stdout "start"
  hFlush stdout
  sendTokens tokens next
  sum <- rootElement cycles tokens this next
  hPutStrLn stdout "end"
  hFlush stdout
  print sum

  putMVar next $! 0
  token3 <- takeMVar this
  
  return (token1 + sum + token3)

newElement this _ = do
  next <- newEmptyMVar
  forkIO (element this next)
  return next

main = do
  args <- getArgs
  next <- newEmptyMVar
  this <- foldM newElement next [2..elements]
  root (read (head args)) (read (head (tail args))) this next

